server/lib/nl/nl_utterance.py

```diff
@@ -0,0 +1,226 @@
+  return places
+
+
+def _save_classifications(classifications: List[NLClassifier]) -> List[Dict]:
+  classifications_dict = []
+  for c in classifications:
+    cdict = {}
+    cdict['type'] = c.type
+
+    if isinstance(c.attributes, ContainedInClassificationAttributes):
+      cip = c.attributes.contained_in_place_type
+      if isinstance(cip, ContainedInPlaceType):
+        cdict['contained_in_place_type'] = cip.value
+      else:
+        cdict['contained_in_place_type'] = cip
+    elif isinstance(c.attributes, RankingClassificationAttributes):
+      cdict['ranking_type'] = c.attributes.ranking_type
+
+    classifications_dict.append(cdict)
+  return classifications_dict
+
+
+def _load_classifications(
+    classifications_dict: List[Dict]) -> List[NLClassifier]:
+  classifications = []
+  for cdict in classifications_dict:
+    attributes = SimpleClassificationAttributes()
+    if 'contained_in_place_type' in cdict:
+      attributes = ContainedInClassificationAttributes(
+          contained_in_place_type=ContainedInPlaceType(
+              cdict['contained_in_place_type']))
+    elif 'ranking_type' in cdict:
+      attributes = RankingClassificationAttributes(
+          ranking_type=[RankingType(r) for r in cdict['ranking_type']],
+          ranking_trigger_words=[])
+    classifications.append(
+        NLClassifier(type=ClassificationType(cdict['type']),
+                     attributes=attributes))
+  return classifications
+
+
+def _save_charts(charts: List[ChartSpec]) -> List[Dict]:
+  charts_dict = []
+  for c in charts:
+    cdict = {}
+    cdict['chart_type'] = c.chart_type
+    cdict['places'] = _save_places(c.places)
+    cdict['svs'] = c.svs
+    cdict['attr'] = c.attr
+    charts_dict.append(cdict)
+  return charts_dict
+
+
+def _load_charts(charts_dict: List[Dict]) -> List[ChartSpec]:
+  charts = []
+  for cdict in charts_dict:
+    charts.append(
+        ChartSpec(chart_type=ChartType(cdict['chart_type']),
+                  places=_load_places(cdict['places']),
+                  svs=cdict['svs'],
+                  attr=cdict['attr'],
+                  utterance=None))
+  return charts
+
+
+# Given the latest Utterance, saves the full list of utterances into a
+# dict.  The latest utterance is in the front.
+def save_utterance(uttr: Utterance) -> List[Dict]:
+  uttr_dicts = []
+  u = uttr
+  cnt = 0
+  while u and cnt < CNTXT_LOOKBACK_LIMIT:
+    udict = {}
+    udict['query'] = u.query
+    udict['query_type'] = u.query_type
+    udict['svs'] = u.svs
+    udict['places'] = _save_places(u.places)
+    udict['classifications'] = _save_classifications(u.classifications)
+    udict['ranked_charts'] = _save_charts(u.rankedCharts)
+    uttr_dicts.append(udict)
+    u = u.prev_utterance
+    cnt += 1
+
+  return uttr_dicts
+
+
+# Given a list of dicts previously saved by `save_utterance()`, loads
+# them into Utterance structures and returns the latest one.
+def load_utterance(uttr_dicts: List[Dict]) -> Utterance:
```

Comment:
i see - this could be a constructor for Utterance from a dict (if Utterance were a regular class, rather than just a dataclass). add a todo for this?